taf/updater/updater_pipeline.py

```diff
@@ -204,20 +204,20 @@ def __init__(
 
         super().__init__(
             steps=[
+                (self.set_existing_repositories, RunMode.UPDATE),
+                (self.check_if_local_repositories_clean, RunMode.UPDATE),
                 (self.clone_remote_and_run_tuf_updater, RunMode.ALL),
                 (self.validate_out_of_band_and_update_type, RunMode.ALL),
-                (self.check_if_local_auth_repo_clean, RunMode.UPDATE),
                 (self.clone_or_fetch_users_auth_repo, RunMode.UPDATE),
                 (self.load_target_repositories, RunMode.ALL),
                 (self.check_if_repositories_on_disk, RunMode.LOCAL_VALIDATION),
-                (self.check_if_local_target_repositories_clean, RunMode.UPDATE),
-                (
-                    self.check_if_local_repositories_contain_unpushed_commits,
-                    RunMode.UPDATE,
-                ),
                 (self.clone_target_repositories_to_temp, RunMode.UPDATE),
                 (self.determine_start_commits, RunMode.ALL),
                 (self.get_targets_data_from_auth_repo, RunMode.ALL),
+                (
+                    self.check_if_local_target_repositories_clean,
+                    RunMode.UPDATE,
+                ),
```

Comment:
This method checks if local target repositories contain unpushed commits? If so, why don't we instead early exit updater if local target contains unpushed commits? Then `self.clone_target_repositories_to_temp` step comes after we've already checked that target repos can be updated. This way we avoid the extra work of cloning target repositories to temp (if the target repo isnt' clean). Am I missing something?

Reply:
In `check_if_local_repositories_clean`, which is run before we try updating the repositories, we check if repositories which are currently on disk are clean and if the most recent published branch (according to the users' authentication repository, before update) contains unpushed commits. This step is run after we fetch the changes and collect the newest data (including a list of branches per repositories which were published). It needs to be run after that information is known.

Reply:
I was thinking we would use git ls-remote. But the command doesn't look like it can list all commits on remote branch :-(.

Reply:
We have a `get_last_remote_commit` method which uses `ls-remote` to find the last remote commit. We could speed things up by using that to check if there are any new commits that should be pulled down. Can be done in a separate PR